src/libcore/macros/mod.rs

```diff
@@ -6,9 +6,13 @@ macro_rules! panic {
     () => (
         $crate::panic!("explicit panic")
     );
-    ($msg:expr) => (
+    ($msg:literal) => (
         $crate::panicking::panic($msg)
     );
+    ($msg:expr) => ({
+        let msg: &str = $msg; // Make sure it is a &str.
```

Comment:
This may cause an issue with temporary lifetimes, see [this](https://stackoverflow.com/questions/48732263/why-is-rusts-assert-eq-implemented-using-a-match).

You can use `match` with `#[allow_internal_unstable(type_ascription)]` to work around this problem:
```rust
match $msg: &str {
    msg => $crate::panic!("{}", msg),
}
```

Reply:
What you linked is about `assert_eq`, which needs this to 1) avoid evaluating the argument(s) twice and/or 2) avoid storing a reference to a temporary. Both are not a problem here, right? We take a copy of the `&str`, and `$msg` is only used once. Am I missing something?

Reply:
Ah, yes, this fails now:

```rust
core::panic!(String::from("hello").as_str());
```

Reply:
The problem with a type ascription like `$crate::panic!("{}", msg: &str)` or your `match` is that it doesn't even allow implicit conversions like `&String -> &str`. With such a type ascription, the following fails:

```rust
core::panic!(&String::from("hello"));
```

with

```
error[E0308]: mismatched types
 --> src/main.rs:2:18
  |
2 |     core::panic!(&String::from("hello"));
  |                  ^^^^^^^^^^^^^^^^^^^^^^ expected `str`, found struct `std::string::String`
  |
  = note: expected reference `&str`
             found reference `&std::string::String
```

Reply:
This seems to work:
```rust
$crate::panic!("{}", $crate::convert::identity::<&str>($msg))
```